### PR TITLE
[proposal] Bundle literals implementation

### LINKED ISSUE #805: [RFC] Bundle Literals
(supersedes #418 with a more detailed proposal of syntax and thoughts on implementation concerns, supersedes #694 with an alternative, I think cleaner, implementation path)

## Motivation
I don't think much needs to be said here. Main use cases would be in testing (for concise and simple testvector specification) and in initializing Bundles with RegInit and WireInit. Maybe also ROMs, but there's a whole other issue for that.

Note: it's possible to construct Bundle-literal-like objects by casting from a UInt with asTypeOf, but specifying anything but all zeros requires knowledge of how Bundles are bit-packed, which is not great. This also requires a Builder context, making it incompatible with testers.

## Proposed syntax
Given a concrete instance of a Bundle type, an autogenerated (by macro annotation) Lit function would return a new literal of that type.

For example, if I have
```scala
@bundle
class MyBundle(val width: Int) extends Bundle {
  val a = Bool()
  val b = UInt(width.W)
}
```

I would be able to do:

```scala
myBundleIo := (new MyBundle(8)).Lit(a=true.B, b=255.U)
```

### Edge cases
It is not necessary to be in a Chisel Builder context to construct a Bundle literal (barring some logic construction being in the Bundle definition, which we have no control over). This allows its use in testers.

Using keyword arguments allows the partial specification of literals. Current proposal is for everything not specified to be treated as `DontCare`.

Bundle literals will take a literal interpretation of `DontCare`, that is, `DontCare` means `DontCare` (and might show up as X-like in simulation) instead of `DoesntExist` (where the previous value on the wire, if any, would not be overwritten).

### Open questions
Should keyword arguments be used? On the plus side, it allows a partial specification, however, it is incompatible with def macros, which are used for source locators. Note that currently, literals do not provide source locators, but this could prevent compatibility with def macros in the future. (current proposal will be to use keyword arguments)

Because Bundles can take parameters, the current proposal requires a literal to be constructed given a Bundle instance (so `Lit` would be a method on an instance, instead of on a companion object). Does it make sense to have a shorthand version in the companion object when no parameters are required? For example, if I defined:
```scala
@bundle
class MyBundle extends Bundle {
  val a = Bool()
  val b = UInt(8.W)
}
```
should `MyBundle.Lit(a=true.B, b=255.U)` be allowed?

### Dealing with LitArg
Literal values are currently stored in the Data subtype objects with an immutable `litArg`. However, this requires literals to be known at instantiation time, which is not possible because fields are instantiated in the Bundle itself, before the literal constructor is run.

Proposal is to move the `litArg` data into LitBinding, which probably makes more sense anyways. Note that this would result in a loss of (internal, within Chisel, not user-visible) type safety, since LitBinding would need to take values for UInt, Bool, and FixedPoint.

### Bindings
Currently, hardware is bound either as a ChildBinding (element of a larger Aggregate) or some kind of leaf binding like LitBinding (indicating an Element is a literal). There are two proposals for binding Bundle literals:

#### Bindings stay top-level
In this approach, LitBinding could contain a map of subelements to their arguments. This works very well with the current structure implementation-wise, however, requires indirection to look up the value of a literal given its Data subtype object. Getting a sub-Bundle literal may also be costly.

It may be possible to have an inconsistent map: containing missing literal values, or extra literal values. Missing literal values could be interpreted as DontCare, or cause an error.

#### Allow intermediate hierarchy nodes to have top-level bindings
The possibilities here are to either have a ChildLitBinding (which has a parent pointer and a literal value) or to drop the parent pointer (ChildBinding) entirely and only have a LitBinding on leaves.

The main issue with the current structure is that Data objects can't be rebound. It might make sense to allow additional bindings: for example, attempting to rebind a Data that has an existing ChildBinding with a new LitBinding will replace the ChildBinding with a ChildLitBinding. Obviously, it's also possible to allow rewriting Bindings, but isn't an elegant solution.

Structural inconsistency is a possibility here, some leaves may be missing their additional LitBinding. In those cases, the system can either raise an error (strict mode, guarding against implementation errors), or treat it as DontCare. But unlike the 'Bindings stay top-level' proposal, it is impossible to have extra literal values.

### Fixing n^2 Bundle instantiation
A somewhat related issue is that Bundle instantiation could be n^2 in clones. That is, each Bundle clone must clone its fields, done recursively. Because of the immutable Data objects abstraction we want to present in Chisel and the possibility of aliasing, even in Bundles (a great example would be the use of genType in multiple fields, perhaps through an `Input(...)` or `Output(...)` binding) I think we want to retain cloning behavior, and I'm not sure if there's a way to avoid the n^2 clones.

A larger refactor that changes the behavior of Bundle-specific `Input(...)`, `Output(...)` or (the proposed) `Field(...)` in an intelligent manner might address this problem, though that's future work. When this becomes a serious bottleneck.

Thoughts, ideas, and suggestions on anything in this proposal are appreciated.

## Issue template

**Type of issue**: other enhancement

**Impact**: API addition (no impact on existing code)

**Development Phase**: proposal

### PR DESCRIPTION
**Pull request by ducky64**
Potential implementation for Bundle literals, that doesn't use macro annotations and has passable syntax.

Example syntax:
```scala
(new MyBundle).Lit(_.a -> 42.U, _.b -> true.B)
```

For nested Bundles:
```scala
// specify an inner bundle with another bundle lit
(new MyOuterBundle).Lit(
  _.a -> (new MyBundle).Lit(_.a -> 42.U, _.b -> true.B)
)

// specify inner bundle fields directly
(new MyOuterBundle).Lit(
  _.a.a -> 42.U, _.a.b -> true.B,
  _.b.c -> false.B, _.b.d -> 255.U
)
```
and more weirdly for anonymous inner Bundles (so probably not recommended syntax?):
```scala
(new MyOuterBundle).Lit(
  b => b.b -> b.b.Lit(_.c -> false.B, _.d -> 255.U)
)
```

The Lit constructor takes in a list of pairs (`->`) from fields to values, each each element being a function from a self-type object (the clone is used internally) to a field. This gives reasonable syntax (basically you have the additional `_.` prefix over each field over what we could do with macro annotations) while being able to compile-time typecheck field accesses (so if you renamed fields, the typechecker would complain about trying to access an nonexistent field).

Drawbacks compared to a macro annotation approach:
- the value type can't be checked at compile time, it accepts any value of type `Data`. With macro annotations, we can (maybe?) generate a function signature based off the field type. Note that compile-time Chisel type checks are not perfect, the dynamic `typeEquivalent` is much more accurate.
- you need the additional magic `_.` prefix for each field.
- `_`  in anonymous functions is a somewhat-flakey operator. This seems to work though. Originally, Lit has the argument signature `(this.type => Data, Data)*` but the `_` doesn't work there.
- technically you could use the `_` operator in the value part (since the argument signature is `(this.type => (Data, Data))*`), but that will exception out inside the Lit constructor, since the Lit constructor calls the function on the newly cloned object (which obviously can't be referenced before).

But not needing a macro annotation (and requiring annotated Bundles, or the dev time to develop the annotation that we haven't had yet) is a nice property...

If we like this API, these improvements are needed:
- more test cases, particularly for error cases
- refactoring of the Lit constructor, while it's not terrible right not it's also not great
- perhaps thoughts on supporting Vec literals, or also Bundle wire initializers (#1005) - but might be a follow-on PR that might need a deeper refactoring

---

**Related issue**: Resolves #805

**Type of change**: other enhancement

**Impact**: API addition (no impact on existing code)

**Development Phase**: proposal

**Release Notes**

### CODE REVIEW

**Comment by chick** [comment]
This seems like a pretty reasonable proposal to me. I think getting them out there with a minimum of boilerplate required of the Bundle definer makes the `_.` a small price to pay. Macro's are nice but make for tough learning of internals. I'd say we should put this on the Friday agenda.

**Comment by edwardcwang** [comment]
Can this work together/is this interoperable with a future `@bundle` annotation?

**Comment by ducky64** [comment]
@edwardcwang potentially: this expects a varargs function of `Data -> Data`, and so it should have a different type signature than what we would generate with a `@bundle` annotation, even if it's also named `Lit` (though we'd want to be really careful to see if there could be potential nasty interactions).

That being said, I think if (and that's a reasonably big if) this goes through, the need for a `@bundle` annotation becomes much less clear. It's a lot of development effort for a small syntactic gain, especially since we already have autoclonetype. The main argument against this PR would be that it's really a bit of a hack, but Chisel has its fair share of syntactic hacks (eg `.elsewhen`, `switch`, the `Module(...)` wrapper).

**Comment by ducky64** [comment]
Resolution from today's meeting was to document partial-initialization and DontCare semantics, and productionize this. The syntax seemed fine, and this is much preferable to the current way of constructing Bundle literals by defining a very low-level `Lit` function.

There was also discussion on naming (`Lit` vs `lit`) and invocation structure (this is very different than existing literal constructors, eg `0.U` - though those may be more converters more like `mySeq.toList`).

**Comment by ducky64** [comment]
This API appears to work in [scastie](https://scastie.scala-lang.org/mYgIVDdaQIOJ4Fso4YHgkg)
```scala
class BundleLitTest {
  val fieldA = 0
  val fieldB = 1
  
  def Lit(elems: (this.type => (Int, Int))*): this.type = {
    for (elem <- elems) {
      println(elem(this))
    }
    this
  }
}

val a = new BundleLitTest
a.Lit(_.fieldA -> 10, _.fieldB -> 11)
```

```
(0,10)
(1,11)
```

**Comment by chick** [comment]
Tested with 2.13 and dotty
[scastie.scala-lang.org](https://scastie.scala-lang.org/pnyoMCs0RFydnQIcqASnoQ)

**Comment by ducky64** [comment]
Ok, I've cleaned up the code and add test cases. Please review now.

Some things that might need more discussion:
- What should `isLit` do for Bundle literals? It currently returns false, partially for backwards compatibility because Bundle literals behave differently than other literals. Should there be a different recommended API?
- Can we have a typeEquivalent definition that the Bundle literal constructor can use to check field assignment? It works for aggregate types. but not for Bits types since the check compares width too.
- Currently, partially initialized Bundle literals have uninitialized fields set to DontCare, and if assigned to another wire, any value on the wire would be overwritten with DontCare. (side note: is there a way to write a test case that a wire is actually DontCare?)
- What should the recommended syntax be for initializing sub-bundles? Deep field specification (potentially breaks abstraction)? Sub-bundle specification (keeps abstraction, but syntax for anonymous inner bundle literals is weird)? Should we support both? Or just one?

Future work would include:
- Vec literals, which also needs discussion on system architecture and API.

**Comment by aswaterman** [comment]
My recommendation is that isLit = elements.forall(_.isLit)--yes, technically an API change, so should consult users.  Same for eventual Vec lits.

The current DontCare scheme sounds like the right way to go.  Assigning a literal to another wire/reg should assign the entire thing--even if some fields are don't-cares.

**Comment by ducky64** [comment]
retest this please

**Comment by ducky64** [comment]
Not sure what's going on with the compile error, it compiles fine for me and I don't see any new style violations. Potentially a transient bug in the build system, or at least that's what the Jenkins log indicates.

**Comment by ucbjrl** [comment]
We're adding some new machines to Jenkins slave network and some of the jobs impose incorrect constraints on their environment. This should be sorted out in the next day or two.

**Comment by chick** [comment]
retest this please

**Comment by chick** [comment]
@ducky64 Looks like this finally made it through the testing. Time to get it merged

**Comment by ducky64** [comment]
One additional thing we can do is to require an experimental import, eg `import chisel3.experimental.BundleLiteralConstructor._` as we've done with the recent testers2 experimental features. Thoughts?

**Comment by ducky64** [comment]
Changed to:
- Require `import chisel3.experimental.BundleLiteralConstructor._` to access Bundle `.Lit`, with the Aggregate class literal constructor changed to `private[chisel3] def _makeLit` (it relies on protected features like `bind` so I can't move the entire thing into an implicit class).
- Remove `Data.selfBind`, which was originally used to support DIY Bundle Literals. Any existing DIY Bundle Literal code will break, but that was never really supported anyways, and I really hope no one else is using the selfBind API.

File: src/test/scala/chiselTests/BundleLiteralSpec.scala [hunk 1]
```diff
@@ -3,47 +3,25 @@
 package chiselTests
 
 import chisel3._
-import chisel3.core.FixedPoint
-import chisel3.experimental.RawModule
 import chisel3.testers.BasicTester
-import org.scalatest._
+import chisel3.experimental.RawModule
+import chisel3.experimental.BundleLiteralConstructor._
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/BundleLiteralSpec.scala]: Suggestion to change `chisel3.experimental.BundleLiteralConstructor` to `chisel3.experimental.BundleLiteral` or `chisel3.experimental.BundleLiterals`

thoughts?

Comment (edwardcwang) [reply src/test/scala/chiselTests/BundleLiteralSpec.scala]: I like the idea of `chisel3.experimental.BundleLiterals`

**Pull request merged**